build-tools-internal/src/integTest/groovy/org/elasticsearch/gradle/internal/SymbolicLinkPreservingTarFuncTest.groovy

```diff
@@ -0,0 +1,163 @@
+  tar.preserveFileTimestamps = ${preserverTimestamp}
+  from fileTree("archiveRoot")
+}
+"""
+        when:
+        gradleRunner("buildTar").build();
+
+        then:
+        assertTar(file("build/distributions/hello-world.tar"), fis -> fis, preserverTimestamp)
+        where:
+        preserverTimestamp << [true, false]
+    }
+
+    private boolean assertTar(final File archive, final FileInputStreamWrapper wrapper, boolean preserveFileTimestamps)
+            throws IOException {
+        try (TarArchiveInputStream tar = new TarArchiveInputStream(wrapper.apply(new FileInputStream(archive)))) {
+            TarArchiveEntry entry = tar.getNextTarEntry();
+            boolean realFolderEntry = false;
+            boolean fileEntry = false;
+            boolean linkToFileEntry = false;
+            boolean linkInFolderEntry = false;
+            boolean linkInFolderLinkToFileEntry = false;
+            boolean linkToRealFolderEntry = false;
+            while (entry != null) {
+                if (entry.getName().equals("real-folder/")) {
+                    assert entry.isDirectory()
+                    realFolderEntry = true
+                } else if (entry.getName().equals("real-folder/file")) {
+                    assert entry.isFile()
+                    fileEntry = true
+                } else if (entry.getName().equals("real-folder/link-to-file")) {
+                    assert entry.isSymbolicLink()
+                    assert normalized(entry.getLinkName()) == "./file"
+                    linkToFileEntry = true
+                } else if (entry.getName().equals("link-in-folder/")) {
+                    assert entry.isDirectory()
+                    linkInFolderEntry = true
+                } else if (entry.getName().equals("link-in-folder/link-to-file")) {
+                    assert entry.isSymbolicLink()
+                    assert normalized(entry.getLinkName()) == "../real-folder/file"
+                    linkInFolderLinkToFileEntry = true
+                } else if (entry.getName().equals("link-to-real-folder")) {
+                    assert entry.isSymbolicLink()
+                    assert normalized(entry.getLinkName()) == "./real-folder"
+                    linkToRealFolderEntry = true
+                } else {
+                    throw new GradleException("unexpected entry [" + entry.getName() + "]")
+                }
+                if (preserveFileTimestamps) {
+                    assert entry.getModTime().getTime() > 0
+                } else {
+                    assert entry.getModTime().getTime() == 0L
+                }
+                entry = tar.getNextTarEntry()
+            }
+            assert realFolderEntry
+            assert fileEntry
+            assert linkToFileEntry
+            assert linkInFolderEntry
+            assert linkInFolderLinkToFileEntry
+            assert linkToRealFolderEntry
+        }
+        true
+    }
+
+
+    interface FileInputStreamWrapper {
```

Comment:
Can't we just use `Function<? super FileInputStream, ? extends InputStream>` instead?

Reply:
fixed